src/index.ts

```diff
@@ -0,0 +1,408 @@
+    const globalProvider = window[GlobalProviders][name];
+    if (!globalProvider) {
+      // Lazily creates a global provider
+      window[GlobalProviders][name] = new Provider({
+        element: document.documentElement,
+        initialState,
+      }).start();
+    } else {
+      // Updates the exiting global provider
+      globalProvider.context = next;
+    }
+  }
+
+  function listen(options: Omit<ListenerOptions<T>, "contextName">) {
+    return new Listener(options).start();
+  }
+  function provide(options: Omit<ProviderOptions<T>, "contextName">) {
+    return new Provider(options).start();
+  }
+  return {
+    name,
+    Provider,
+    Listener,
+    listen,
+    provide,
+    provideGlobally,
+  };
+}
+
+export class ContextListener<T> {
+  resolvePromise?: Resolve<T>;
+  _status: ListenerConnectionStatus = ListenerConnectionStatus.INITIAL;
+  _interval?: NodeJS.Timeout;
+  options: ListenerOptions<T>;
+
+  constructor(options: ListenerOptions<T>) {
+    this.options = options;
+  }
+
+  set status(next: ListenerConnectionStatus) {
+    this._status = next;
+    this.options.onStatus && this.options.onStatus(next);
+  }
+  get status() {
+    return this._status;
+  }
+
+  /* called by provider */
+  onChange = (context: T) => {
+    this.options.onChange && this.options.onChange(context);
+  };
+
+  /* called by provider */
+  onConnect = async (context: T) => {
+    this.status = ListenerConnectionStatus.CONNECTED;
+    this.options.onChange && this.options.onChange(context);
+    return new Promise((resolve) => {
+      this.resolvePromise = resolve;
+    });
+  };
+
+  /* called by provider */
+  onDisconnect = () => {
+    this._status = ListenerConnectionStatus.CONNECTING;
+    this.start();
+  };
+
+  /**
+   * Starts this Listener
+   *
+   * Dispatches an event in the dom, which bubbles up to the nearest ancestor that can handle it.
+   * If no ancestor handles the event, then will retry a few times and error on failure.
+   *
+   */
+  start() {
+    let attempts = 0;
+    this.status = ListenerConnectionStatus.CONNECTING;
+    const getStatus = () => this.status;
+    const maxAttempts = get(this.options.attempts) || ATTEMPTS;
+    const pollingMs = get(this.options.pollingMs) || POLLING;
+    const tryConnect = () => {
+      if (getStatus() !== ListenerConnectionStatus.CONNECTED) {
+        const event = createEvent(this.options.contextName, {
+          onConnect: this.onConnect,
+          onChange: this.onChange,
+          onDisconnect: this.onDisconnect,
+        });
+        get(this.options.element).dispatchEvent(event);
+
+        if (getStatus() !== ListenerConnectionStatus.CONNECTED) {
+          attempts++;
+          if (attempts >= maxAttempts) {
+            this._interval && clearInterval(this._interval);
+
+            this.status = ListenerConnectionStatus.TIMEOUT;
+            // throw new Error(`Gave up trying to connect to provider`);
```

Comment:
Was no longer throwing an exception on connection failure intentional?

Reply:
This was causing tests to fail because `start` would throw an exception, but maybe should continue to throw? It was mostly beneficial because it logs an error to the console, but I'm on the fence to it's continued utility.